Add mobile style rules & plugin API

kellyi opened the pull request:
## Overview

This PR adds some style rules & interfaces to enable the app to show a mobile version on devices sized up to iPhone 6/7/8s-sized widths. To accomplish this, I added a footer component with buttons which only display when the media query for a phone-sized device matches, along with some logic to handle toggling in the app.

The app also now adds two methods for plugins in single app mode: `toggleMobileMap` and `toggleMobileContent`. These toggle each of those on, respectively, and change the state such that the buttons & toggle icon are adjusted in tandem. I added some error messages & noops to handle cases in which the methods don't exist for whatever reason, but left them exposed & callable in non-mobile versions of the app. (We can adjust the names to something which doesn't include "mobile", btw, if we decide to keep them accessible for all screen sizes).

Connects #1002
Connects #1003 

## Demo

I tested this out with a few iPhone simulators and on a 6s; we also took a look at an Android version on @rajadain's phone

Here's iPhone SE

![iphonese](https://user-images.githubusercontent.com/4165523/31511210-fcb00240-af54-11e7-96ec-c1ab1ce5f3a1.gif)

And 7S:

![iphone7s](https://user-images.githubusercontent.com/4165523/31511291-4127f248-af55-11e7-8abe-1f307e6601ae.gif)

## Notes

- Still working through some CSS stuff wrt to the search bar: I think the ideal behavior is that when it's expanded, it occupies the full breadth of the header from left margin to the search icon -- and also covers up the app header title and "About" button.

- I think that some of the app's external CSS rules warrant some changes to make them rely more on viewport size definitions and less on percentages. We can spin that off to another card.

## Testing
- get and rebuild this branch
- visit the app in the browser in single plugin mode in a regular browser-sized viewport and verify that it works as before
- change the viewport to a mobile sized browser then hard refresh and verify that you see mobile mode and that everything is functioning
- check a few mobile viewport sizes to verify that the app appears correctly in each of them

kellyi commented:
About to update the demo gif btw.

caseycesari commented on src/GeositeFramework/js/SidebarToggle.js:
```diff
@@ -47,20 +47,84 @@
+            this.mobileMapButton.on('click', this.showMobileMap);
+            this.mobileContentButton.on('click', this.showMobileContent);
+
+            N.app.toggleMobileMap = this.showMobileMap;
```
I think I prefer the internal (`show...`) name of these functions over the public (`toggle...`) name of these functions, since they will always show and not toggle the appropriate content. Is that right?

kellyi replied:
Yeah -- I'll change these throughout to `show...`.

caseycesari commented on src/GeositeFramework/js/SidebarToggle.js:
```diff
@@ -47,20 +47,84 @@
+
+        showMobileContent: function() {
+            var singlePlugin = this.viewModel.get('plugins').at(0);
+            singlePlugin.select();
```
Can this be omitted? `select` may have side-effects, like reseting controls and options made by the user to the plugin.

kellyi replied:
Let me check on this. My initial thought was that we'd have the buttons behave the same way as the sidebar collapse tab chevron button works in the desktop single web app: it calls `select` and `deselect` when the plugin's expanded and collapsed, respectively.

The idea here was that I wanted to keep them in harmony so now both the `TogglePlugin` view and the `MobileTogglePlugin` view listen for & change on the same model events. 

However, I think you're suggesting that we shouldn't ever deselect the plugin in mobile view -- and instead just show and hide it?

caseycesari replied:
That makes sense, but I think we don't actually want it to be in harmony with that button because we want to the plugin to maintain all of it's state in this case. Changes in `select/deselect` don't necessarily reset or lose the plugin state, but I think it can if a plugin chooses to do so. 

Does that sound right @mmcfarland?

kellyi replied:
I updated these to use `show`/`hide` instead of select & deselect.

kellyi replied:
Oof, just realized I updated the wrong set. Pushing another fixup.

caseycesari commented on src/GeositeFramework/js/SidebarToggle.js:
```diff
@@ -47,20 +47,84 @@
+            this.showMobileContent = this.showMobileContent.bind(this);
+
+            this.mobileMapButton.on('click', this.showMobileMap);
+            this.mobileContentButton.on('click', this.showMobileContent);
```
If possible, consider assigning these events using the Backbone [event delegation](http://backbonejs.org/#View-events) style.

caseycesari commented on src/GeositeFramework/js/SidebarToggle.js:
```diff
@@ -47,20 +47,84 @@
+    N.views.MobileTogglePlugin = Backbone.View.extend({
+        initialize: function(options) {
+            this.viewModel = options.viewModel;
+            this.mobileContentButton = $('#single-plugin-toggle-content-button');
```
This won't have a big effect, but consider searching the view's `$el` for these elements.

caseycesari commented on src/GeositeFramework/js/Plugin.js:
```diff
@@ -97,6 +99,28 @@ require(['use!Geosite',
 
+        function toggleMobileMap() {
+            if (!this.singlePluginMode) {
+                console.error('Programmatic map toggle only available in single plugin mode');
```
This is just my preference, but I don't feel that we need to add a message to the console is these cases. What do you think @mmcfarland?

kellyi replied:
Going to drop these error messages; without them, these functions are small enough that we can just make the assignments inline in the initializer:

```js
showMobileMap = (N.app.singlePluginMode && N.app.showMobileMap) ?
    N.app.showMobileMap : _.noop(),
showMobileContent = (N.app.singlePluginMode && N.app.showMobileContent) ?
    N.app.showMobileContent : _.noop();
```

caseycesari commented on src/GeositeFramework/css/main.css:
```diff
@@ -2120,3 +2120,95 @@ body.x-body {
+}
+
+@media screen and (max-device-width: 736px) {
+    #toggle-plugin-container {
```
Are the `!important` flags used in this block required? I thought that the rules defined within the media query would take precedent over other rules when the media query is active.

kellyi replied:
Some of these do seem to be required, but I'll go through and check to make sure we keep only the minimal number of them.

kellyi replied:
I pared these down to what seem to be the minimal required set.

caseycesari commented:
The about pane should be displayed above the toggle buttons since they will have no effect while the about pane is open:

![image](https://user-images.githubusercontent.com/1042475/31460315-0364844c-ae94-11e7-80c4-db815d4c5f10.png)

kellyi commented:
> I tested this on my phone in mobile Safari, and I'm not able to scroll to the toggle buttons.

Ah, this is due to changing the app height from 100% to 100vh -- forgot that mobile Safari counts the bit beneath the navigation bar as part of the viewport.

kellyi commented:
ac84e90 addresses the code comments, makes some CSS fixes, and also adjusts the geocoder search bar such that we now hide the app title & subtitle when the search bar's expanded & show it when it's collapsed.

I updated the demo gifs in the first comment above, too.

One additional testing instruction is that in during single app mobile mode, you should be able to call the following functions in the console & see the desired effects:

```
Geosite.app.showMobileMap();
Geosite.app.showMobileContent();
```

`showMobileMap` should show the map & toggle the buttons such that the map button's highlighted; `showMobileContent` should show the content pane & toggle the buttons such that the content button's highlighted.

caseycesari commented:
Oh, also the button focus color issue!

kellyi commented:
814bc7a handles both hiding the mobile button footer for non-mobile devices & also dropping the green-ish focus class from the toggle buttons

kellyi commented:
Great! I just squashed this, then rebased it on develop & tested it out again. Everything's still working as expected, so I'm going to merge this in.

Thanks for all your help with this!
